Tests/NIOPosixTests/IPv4Header.swift

```diff
@@ -0,0 +1,277 @@
+            destinationIpAddress
+        ) = buffer.readMultipleIntegers(as: (
+            UInt8,
+            UInt8,
+            UInt16,
+            UInt16,
+            UInt16,
+            UInt8,
+            UInt8,
+            UInt16,
+            UInt32,
+            UInt32
+        ).self) else { return nil }
+        #endif
+        self.versionAndIhl = versionAndIhl
+        self.dscpAndEcn = dscpAndEcn
+        self.totalLength = totalLength
+        self.identification = identification
+        self.flagsAndFragmentOffset = flagsAndFragmentOffset
+        self.timeToLive = timeToLive
+        self.`protocol` = .init(rawValue: `protocol`)
+        self.headerChecksum = headerChecksum
+        self.sourceIpAddress = .init(rawValue: sourceIpAddress)
+        self.destinationIpAddress = .init(rawValue: destinationIpAddress)
+    }
+    init() {
+        self.versionAndIhl = 0
+        self.dscpAndEcn = 0
+        self.totalLength = 0
+        self.identification = 0
+        self.flagsAndFragmentOffset = 0
+        self.timeToLive = 0
+        self.`protocol` = .init(rawValue: 0)
+        self.headerChecksum = 0
+        self.sourceIpAddress = .init(rawValue: 0)
+        self.destinationIpAddress = .init(rawValue: 0)
+    }
+    
+    mutating func setChecksum() {
+        self.headerChecksum = ~[
+            UInt16(versionAndIhl) << 8 | UInt16(dscpAndEcn),
+            totalLength,
+            identification,
+            flagsAndFragmentOffset,
+            UInt16(timeToLive) << 8 | UInt16(`protocol`.rawValue),
+            UInt16(sourceIpAddress.rawValue >> 16),
+            UInt16(sourceIpAddress.rawValue & 0b0000_0000_0000_0000_1111_1111_1111_1111),
+            UInt16(destinationIpAddress.rawValue >> 16),
+            UInt16(destinationIpAddress.rawValue & 0b0000_0000_0000_0000_1111_1111_1111_1111),
+        ].reduce(UInt16(0), onesComplementAdd)
+        
+        assert(isValidChecksum())
+    }
+    
+    func isValidChecksum() -> Bool {
+        let sum = ~[
+            UInt16(versionAndIhl) << 8 | UInt16(dscpAndEcn),
+            totalLength,
+            identification,
+            flagsAndFragmentOffset,
+            UInt16(timeToLive) << 8 | UInt16(`protocol`.rawValue),
+            headerChecksum,
+            UInt16(sourceIpAddress.rawValue >> 16),
+            UInt16(sourceIpAddress.rawValue & 0b0000_0000_0000_0000_1111_1111_1111_1111),
+            UInt16(destinationIpAddress.rawValue >> 16),
+            UInt16(destinationIpAddress.rawValue & 0b0000_0000_0000_0000_1111_1111_1111_1111),
+        ].reduce(UInt16(0), onesComplementAdd)
+        return sum == 0
+    }
+    
+    func write(to buffer: inout ByteBuffer) {
+        assert({
+            var buffer = ByteBuffer()
+            self._write(to: &buffer)
+            let newValue = Self(buffer: &buffer)
+            return self == newValue
+        }())
+        self._write(to: &buffer)
+    }
+    
+    private func _write(to buffer: inout ByteBuffer) {
```

Comment:
normally we write these as

```
extension ByteBuffer {
   private func writeTCP4Packet...
```

Reply:
Done in https://github.com/apple/swift-nio/pull/2320/commits/20ed3802642a475b760beb33d8a121430f0224a6